### ISSUE TITLE
Massive Performance Issue When Using Redis

### ISSUE THREAD

**Issue by toddca** [nopSolutions/nopCommerce#3875]
nopCommerce version: 4.2

Steps to reproduce the problem: 
1. Setup Nop to use Redis Caching
2. Using the default theme and sample data monitor the number of calls which are made from nopCommerce server to Redis.

There are a ton of calls to Redis from NopCommerce, its very chatty to the point you cannot run any amount of load without overloading the client queue and timeout issues start popping.

I added logging to the RedisCacheManager just before we call into Redis. I then started the webapp and made a request to the home page and 2 category pages. Here are the stats for just those 3 page requests:
> Total Redis Calls: 1,187
> Total Calls to KeyExists: 705
> Total Calls to StringGet: 373
> Total Calls to StringSet: 109
> Calls to KeyExists('Nop.lsr.all-1'): 215
> Calls to KeyExists('Nop.setting.all'): 232

This is attached as all.log
[all.log](https://github.com/nopSolutions/nopCommerce/files/3198362/all.log)

Focusing on one request to a catalog page for which I have already requested there are 186 calls to Redis - 1/3 of those are to KeyExists('Nop.lsr.all-1'). Basically any call to get a resource string will call into GetAllResourceValues() which on line 266 calls _cacheManager.IsSet().

This is attached as 1request.txt
[1request.txt](https://github.com/nopSolutions/nopCommerce/files/3198365/1request.txt)

Running the site under a 25 user load making requests with zero think time you get a 5% error rate due to RedisTimeoutException

Requests: 30,241
Errors: 1,608 (5%)
RPS: 100.8
Avg Response Time: 646.4ms

Keeping everything the same and just turning off Redis the same test produces
Requests: 93,649
Errors: 7 (0%)
RPS: 312.2
Avg Response Time: 232.4ms

The bottom line - without Redis the server is over 300% faster and can do 3x the traffic and with much fewer errors. At this point I do not see the advantage of using Redis for caching and would advise anyone to not use it.

**Comment by skoshelev** [comment]
We know about redis performance problems, unfortunately we should to use the sinc requests to the redis. As for the problem with timeouts, the problem is solved by increasing the waiting time in the connection string. In the future we will rewrite everything to async calls (#2287), which will improve performance.

Also see the #3797 tiket

Closed #3875

**Comment by jspuij** [comment]
I feel there are some other things with redis as well. We have been able to improve performance of the front end significantly by adding https://github.com/johnnycardy/StackRedis.L1 as an in memory cache in between Redis and NopCommerce.

Redis is currently used in the code base as a glorified dictionary. But every key acces is a network request and slows things down. Even with an async code base this will still be horrible. So we added a memory cache that somebody else already wrote: StackRedis.L1.
For now StackRedis.L1 does it's job for as as subsequent key requests will happen in memory and are much faster. Yet we retain global invalidation of keys and prefixes as it uses redis events to keep the memory caches in sync.

Then another issue. Whenever a Category, Product, Setting or whatever is saved through its service, the EventPublisher will call 4 consumers: ModelCacheEventConsumer (1 regular, 1 admin), DiscountEventConsumer, PriceEventConsumer. And these pretty much delete all keys for any entity type.

This leads to a cascade of Redis traffic. We counted **300.000** network requests to the redis server to obtain and delete keys by saving the general settings page.

This leads to a 30 second timeout on our load balancer, on a hyperconverged cluster with 56Gb fiber
interconnects. I don't want to know how much time this would take on a "normal" network.

I've got a feeling that the CacheEventConsumers could be smarter by filtering based on the type of event received. Also there is no reason to update a setting or fire the event handlers when a property has not changed, so old / newvalue should be compared in the controllers.

We did a proof of concept on the admin setting controller and were able to improve performance 100x. 

As it's a lot of work to update all controllers / consumers, I'm not sure how to tackle this going forward. Create a PR? Anyone on the core team able to comment?

**Comment by erikest** [comment]
We've come to the same conclusion as others - we're seeing major issues even with connectTimeout=30000 and very little traffic on our staging server, so we're having to abandon redis for now and circle back after reimplementing it as async and two-layer with memory cache.  There's already a memory cache implementation in Nop, it's just one or the other with the redis cache, instead of memory cache backed by redis.  We will look at StackRedis.L1.

agree with @jspuij, that performance could be improved dramatically by reducing cache chatter.  I see a couple hundred per request.  The current 'performance' hack that _is_ in there is a per-request cache to prevent multiple redis calls for the same key on the same request, but even with that, there's way too much talking.

**Comment by jspuij** [comment]
for anyone interested:
https://github.com/BirdseyeviewNL/nopCommerce/tree/redis-improvements

Increasing page size and inserting StackRedis.L1 has made a difference. For other changes that made a difference for us:

https://github.com/BirdseyeviewNL/nopCommerce/tree/develop-with-local-fixes

**Comment by sinaislam** [comment]
> for anyone interested:
> https://github.com/BirdseyeviewNL/nopCommerce/tree/redis-improvements
> 
> Increasing page size and inserting StackRedis.L1 has made a difference. For other changes that made a difference to us:
> 
> https://github.com/BirdseyeviewNL/nopCommerce/tree/develop-with-local-fixes
> 
> As a general note on the use of large page sizes with Redis:
> 
> As Redis is singlethreaded, large page sizes will block the Redis server for other requests. Other issue: network packets will become large and fragmented.

Do you really change anything or just shared your thought? I download your branch and do not see any change from nopCommerce4.2. Did you implement the StackRedis.L1 at your branch? Was that helpfull?

**Comment by colinwphillips** [comment]
> > for anyone interested:
> > https://github.com/BirdseyeviewNL/nopCommerce/tree/redis-improvements
> > Increasing page size and inserting StackRedis.L1 has made a difference. For other changes that made a difference to us:
> > https://github.com/BirdseyeviewNL/nopCommerce/tree/develop-with-local-fixes
> > As a general note on the use of large page sizes with Redis:
> > As Redis is singlethreaded, large page sizes will block the Redis server for other requests. Other issue: network packets will become large and fragmented.
> 
> Do you really change anything or just shared your thought? I download your branch and do not see any change from nopCommerce4.2. Did you implement the StackRedis.L1 at your branch? Was that helpfull?

https://github.com/BirdseyeviewNL/nopCommerce/commit/e9eaa736f69ba568f66874009d557ed7afed61ae

The shared repo was very useful and easy to implement locally. We’ve seen tremendous improvement and no longer see any timeouts whatsoever, down from thousands per day.

**Comment by jspuij** [comment]
> > for anyone interested:
> > https://github.com/BirdseyeviewNL/nopCommerce/tree/redis-improvements
> > Increasing page size and inserting StackRedis.L1 has made a difference. For other changes that made a difference to us:
> > https://github.com/BirdseyeviewNL/nopCommerce/tree/develop-with-local-fixes
> > As a general note on the use of large page sizes with Redis:
> > As Redis is singlethreaded, large page sizes will block the Redis server for other requests. Other issue: network packets will become large and fragmented.
> 
> Do you really change anything or just shared your thought? I download your branch and do not see any change from nopCommerce4.2. Did you implement the StackRedis.L1 at your branch? Was that helpfull?

checkout either the redis-improvements branch or the develop-with-local-fixes branch. I made no changes in the master branch of my fork. It works for me, and apparently for @colinwphillips , but YMMV. I have not made into a full pull request yet because Nopcommerce is a very large system and I feel i'm currently not familiar enough with the code base to make fundamental changes to the caching mechanism. I was able to do a quick analysis, fix some stuff and alleviate our worst issues. I'm willing to invest that time only if it's appreciated. Not going to invest a weeks work for it not te be picked up afterwards. I have a PR from august, fixing somebody else's issue, that has not been discussed or looked at. So I'm holding of putting more effort into it until there is some guidance on how to proceed.

**Comment by sinaislam** [comment]
> > > for anyone interested:
> > > https://github.com/BirdseyeviewNL/nopCommerce/tree/redis-improvements
> > > Increasing page size and inserting StackRedis.L1 has made a difference. For other changes that made a difference to us:
> > > https://github.com/BirdseyeviewNL/nopCommerce/tree/develop-with-local-fixes
> > > As a general note on the use of large page sizes with Redis:
> > > As Redis is singlethreaded, large page sizes will block the Redis server for other requests. Other issue: network packets will become large and fragmented.
> > 
> > 
> > Do you really change anything or just shared your thought? I download your branch and do not see any change from nopCommerce4.2. Did you implement the StackRedis.L1 at your branch? Was that helpfull?
> 
> checkout either the redis-improvements branch or the develop-with-local-fixes branch. I made no changes in the master branch of my fork. It works for me, and apparently for @colinwphillips , but YMMV. I have not made into a full pull request yet because Nopcommerce is a very large system and I feel i'm currently not familiar enough with the code base to make fundamental changes to the caching mechanism. I was able to do a quick analysis, fix some stuff and alleviate our worst issues. I'm willing to invest that time only if it's appreciated. Not going to invest a weeks work for it not te be picked up afterwards. I have a PR from august, fixing somebody else's issue, that has not been discussed or looked at. So I'm holding of putting more effort into it until there is some guidance on how to proceed.

My Bad. Sorry actually I did not notice the branch. Thank you. I need to implement it for nopCommerce 4.1. I will must try your branch.

**Comment by Durkee** [comment]
> for anyone interested:
> https://github.com/BirdseyeviewNL/nopCommerce/tree/redis-improvements
> 
> Increasing page size and inserting StackRedis.L1 has made a difference. For other changes that made a difference to us:
> 
> https://github.com/BirdseyeviewNL/nopCommerce/tree/develop-with-local-fixes
> 
> As a general note on the use of large page sizes with Redis:
> 
> As Redis is singlethreaded, large page sizes will block the Redis server for other requests. Other issue: network packets will become large and fragmented.

Hello.  We are having throughput issues and would like to try our your branch.  Can you contact me to discuss some of our deployment strategies?   sedREMOVE@storis.com  
Appreciate it,
Steve

**Comment by MDRashedKhanMenon** [comment]
my client is facing serious slowness on nop 4.2 .Is this issue will remain unsolved?

**Comment by sinanopcommerce** [comment]
> my client is facing serious slowness on nop 4.2 .Is this issue will remain unsolved?

Did you check out the branch? You have to change the code according to the branch. It seems his branch is perfect.
